### ISSUE TITLE
Single site popgen stats following tskit

### ISSUE THREAD

**Issue by jeromekelleher** [pystatgen/sgkit#94]
It would be useful to have a set of basic single site popgen statistics implemented to help guide overall API design and provide us with easy benchmarks to run. Having put a fair bit of thought and effort into this recently for [this paper](https://www.genetics.org/content/215/3/779), I'm hoping that we can borrow some of the API design and semantic choices from the [tskit stats API](https://tskit.readthedocs.io/en/latest/stats.html).

My suggestion would be to implement sgkit functions that correspond to:

- [diversity](https://tskit.readthedocs.io/en/latest/python-api.html#tskit.TreeSequence.diversity)
- [divergence](https://tskit.readthedocs.io/en/latest/python-api.html#tskit.TreeSequence.divergence)
- [Fst](https://tskit.readthedocs.io/en/latest/python-api.html#tskit.TreeSequence.Fst)
- [Tajima's D](https://tskit.readthedocs.io/en/latest/python-api.html#tskit.TreeSequence.Tajimas_D)
- [Patterson's F4](https://tskit.readthedocs.io/en/latest/python-api.html#tskit.TreeSequence.f4)

This would be a good overview of the basic popgen stats, and I think give us a good sample of code to generalise from in terms of coming up with multiple dispatch backends, etc.

Obviously I'm biased in suggesting that we follow the tskit definitions closely, but we did spend a long time thinking through how best to arrange the API and semantics, and the implementations are *very* thoroughly tested. It would also make our lives a bit easier here if we don't have so many choices to make, but of course that's up for discussion. A big advantage is that it would make testing the implementations straightforward, since we could simply generate some simulated test data using [msprime](https://msprime.readthedocs.io/en/latest/), and verify that we get the same answer as tskit.

What do we think of this as a general strategy for putting PopGen stats into the library? I'm not saying we have to slavishly copy the function and parameter names etc, but rather use them as a guide.

**Comment by eric-czech** [comment]
> I think give us a good sample of code to generalise from in terms of coming up with multiple dispatch backends, etc.

These generalizations would include distributed processing in your mind right?

> since we could simply generate some simulated test data using msprime, and verify that we get the same answer as tskit

Sounds great to me, it's a similar approach to what I took with https://github.com/pystatgen/sgkit/issues/31 and scikit-alelle for verification so we'd align nicely.

> What do we think of this as a general strategy for putting PopGen stats into the library?

So I'm understanding you correctly, you do mean something like mapping the tskit API onto the similar statistics in scikit-allele that run over arrays and not trees correct?  

No objections here in sticking to those semantics as closely as possible.  For the API, I'm assuming a possible change would be making the functions operate on a data structure rather than a data structure that exposes the functions.  Is there any important state saved internally within those classes or is that kind of change as easy as it seems?

**Comment by jeromekelleher** [comment]
>> I think give us a good sample of code to generalise from in terms of coming up with multiple dispatch backends, etc.
> These generalizations would include distributed processing in your mind right?

Sure. Just to make sure we're on the same page, I'm suggesting we start out with direct xarray implementations like you've done for allele counting, and then once these are merged we can stand back and look at all ~10 different things we've implemented (including the operations you've been working on), and see what the general requirements for dispatching to CUDA, dask, etc are. I think it's very hard to predict what API requirements are until we have a reasonable number of concrete API endpoints that we know we want to support. Also, it'll be good to have benchmarks for a number of different operations, and these stats will be good for that.

> So I'm understanding you correctly, you do mean something like mapping the tskit API onto the similar statistics in scikit-allele that run over arrays and not trees correct?

Basically, yes. There's some details to work out about what the functions should operate on (i.e., datasets or allele counts, or a mixture), but I'll come back here with some proposals in a bit.

**Comment by jeromekelleher** [comment]
Here's a first pass at what this might look like:
```python
import msprime
import sgkit
import numpy as np


def diversity(ac):
    an = ac.sum(axis=1)
    n_pairs = an * (an - 1) / 2
    n_same = (ac * (ac - 1) / 2).sum(axis=1)
    n_diff = n_pairs - n_same
    # Let's ignore missing data and division by zero for now.
    pi = n_diff / n_pairs
    # Because we're not providing any arguments on windowing, etc,
    # we return the total over the whole region. Maybe this isn't
    # the behaviour we want, but it's a starting point. Note that
    # this is different to the tskit default behaviour where we
    # normalise by the size of windows so that results
    # in different windows are comparable. However, we don't have
    # any information about the overall length of the sequence here
    # so we can't normalise by it.
    return pi.sum()


def divergence(ac1, ac2):

    an1 = ac1.sum(axis=1)
    an2 = ac2.sum(axis=1)

    n_pairs = an1 * an2
    n_same = (ac1 * ac2).sum(axis=1)
    n_diff = n_pairs - n_same
    # Ignore missing data and division by zero for now.
    div = n_diff / n_pairs
    return div.sum()

def ts_to_dataset(ts, samples=None):
    """
    Convert the specified tskit tree sequence into an sgkit dataset.

    Note this just generates haploids for now. With msprime 1.0, we'll be
    able to generate diploid/whatever-ploid individuals easily.
    """
    if samples is None:
        samples = ts.samples()
    tables = ts.dump_tables()
    alleles = []
    genotypes = []
    for var in ts.variants(samples=samples):
        alleles.append(var.alleles)
        genotypes.append(var.genotypes)
    alleles = np.array(alleles).astype("S")
    genotypes = np.expand_dims(genotypes, axis=2)

    df = sgkit.create_genotype_call_dataset(
        variant_contig_names=["1"],
        variant_contig=np.zeros(len(tables.sites), dtype=int),
        variant_position=tables.sites.position.astype(int),
        variant_alleles=alleles,
        sample_id=np.array([f"tsk_{u}" for u in samples]).astype("U"),
        call_genotype=genotypes,
    )
    return df


ts = msprime.simulate(20, length=100, mutation_rate=0.05, random_seed=42)

df = ts_to_dataset(ts)
ac = sgkit.count_alleles(df).compute()
pi = diversity(ac)
pi_tsk = ts.diversity(span_normalise=False)
assert np.allclose(pi, pi_tsk)


subset_1 = ts.samples()[: ts.num_samples // 2]
subset_2 = ts.samples()[ts.num_samples // 2 :]
df1 = ts_to_dataset(ts, subset_1)
df2 = ts_to_dataset(ts, subset_2)
ac1 = sgkit.count_alleles(df1).compute()
ac2 = sgkit.count_alleles(df2).compute()

div_ts = ts.divergence([subset_1, subset_2], span_normalise=False)
div = divergence(ac1, ac2)
assert np.allclose(div_ts, div)
```

The code is lifted almost directly from [scikit-allel](https://scikit-allel.readthedocs.io/en/stable/_modules/allel/stats/diversity.html#mean_pairwise_difference), but modified to use xarray methods rather than numpy functions.

Interesting questions start popping up almost immediately:

- Should we be annotating the allele counts object with information about the dataset? For example, if we want to do windowing here and we consider the default window to be the whole sequence, then we need to know how long this is to normalise. There's no way to know this if we just pass in raw allele counts.
- It's a bit clunky having to compute the allele counts up front - maybe in a higher-level API we'd do this under the hood, and cache any allele counts on the object in question?
- Subset handling is also pretty clunky - do we want to have a separate dataset for each of the subsets when we're working with things like divergence, or would it be better to take the approach that tskit has where we specify the list of samples we are interested explicitly.

I don't think we need to answer these questions immediately though - we can probably just get on with implementing these basic operations and adding tests, and then circle back to ask these API issues when we have the implementations done.

Any objections to adding functions like these for the 5 stats I identified above?

**Comment by eric-czech** [comment]
> Any objections to adding functions like these for the 5 stats I identified above?

Nope not at all!  Or at least certainly not from me -- thanks for the examples.

> Interesting questions start popping up almost immediately:

I'm not at all saying these are the right ways to do them, but I had similar questions arise and here are the solutions I had been adopting:

> Should we be annotating the allele counts object with information about the dataset? 

If you needed positions for windowing, you could just assume allele counts are one variable provided along with any others.  Something like this?

```python
ds = ts_to_dataset(ts)
ds['variant_allele_count'] = sgkit.count_alleles(ds)
# Or ds = ds.merge(sgkit.count_alleles(ds))
pi = diversity(ds)
```

> It's a bit clunky having to compute the allele counts up front

I had similar situations come up with regression and HWE tests.  The approach I had been taking was something like:

```python
def diversity(ds: Dataset, allele_counts: Hashable='variant_allele_counts'):
  if allele_counts not in ds:
    ds[allele_counts] = sgkit.count_alleles(ds)
  ... 
```

> Subset handling is also pretty clunky

I think a more generic way to subset could be:

```python
from sgkit import count_alleles
from sgkit.testing import simulate_genotype_call_dataset

ds = simulate_genotype_call_dataset(n_variant=100, n_sample=20, seed=0)
ds['sample_population'] = np.repeat([0, 1], ds.dims['samples'] // 2)
ac = ds.groupby('sample_population').map(lambda ds: count_alleles(ds))
ac
```
![Screen Shot 2020-08-06 at 10 46 08 AM](https://user-images.githubusercontent.com/6130352/89545997-0e39e280-d7d2-11ea-9bdc-78a45dad0cda.png)

And then:

```python
divergence(ac.isel(sample_population=0), ac.isel(sample_population=1))
```

**Comment by jeromekelleher** [comment]
Awesome, thanks for this @eric-czech, lots to think about!

@daletovar - how does the above look to you?

**Comment by daletovar** [comment]
> I'm suggesting we start out with direct xarray implementations like you've done for allele counting, and then once these are merged we can stand back and look at all ~10 different things we've implemented (including the operations you've been working on), and see what the general requirements for dispatching to CUDA, dask, etc are.

I agree. I think this is the best way forward. I can work on adding these.

**Comment by jeromekelleher** [comment]
> I agree. I think this is the best way forward. I can work on adding these.

@daletovar - excellent! Please feel free to ping me on any technical details. I'm a big fan of draft PRs, so please do open one up as soon as you have something halfway working.

**Comment by daletovar** [comment]
@jeromekelleher, thanks! I'll write something up.

**Comment by tomwhite** [comment]
@jeromekelleher do you think we can close this now we have the initial set of popgen stats?
